index.js

```diff
@@ -0,0 +1,116 @@
+const lazyloadOptions = {
+  rootElement: null,
+  rootMargin: `0px 0px ${window.innerHeight / 2}px 0px`,
+  treshold: 1,
+  lazyloadClass: 'loaded',
+  nativeLazyloadEnabled: 'loading' in HTMLImageElement.prototype,
+
+  getOptions() {
+    return {
+      root: this.rootElement,
+      rootMargin: this.rootMargin,
+      threshold: this.treshold
+    };
+  }
+};
+
+let lazyObserver = null;
+let resizeListener = null;
+let elements = [];
+let images = [];
+
+const setSource = element => {
+  const { src, srcset, wait } = element.dataset;
+
+  if (src) {
+    element.src = src;
+  }
+
+  if (srcset) {
+    element.srcset = srcset;
+  }
+
+  if (wait) {
+    element.onload = () => element.classList.add(lazyloadOptions.lazyloadClass);
+  }
+
+  element.removeAttribute('data-lazy');
+};
+
+const loadElements = () => {
+  elements = Array.from(document.querySelectorAll('audio[data-lazy], video[data-lazy], iframe[data-lazy], script[data-lazy]'));
+  images = Array.from(document.querySelectorAll('img[data-lazy]'));
```

Comment:
These selectors are a bit odd to me. We're repeating `[data-lazy]` for each one. Also, not accounting for `divs` or `spans` with background images. 

Maybe grab all and if the native lazy load is available filter images out or feel free to take any other approach you want to.